### ISSUE TITLE
BUG in graph_actions.py

### ISSUE THREAD

**Issue by david-liu** [tensorflow/tensorflow#4203]
Hello I am using `tf.contrib.learn.Estimator` to train my model, and size of validation sample is small,(validation sample size=32, eval_batch_size=16)

I have created an EvaluationMonitor to evaluate the validation sample

```
class EvaluationMonitor(tf.contrib.learn.monitors.EveryN):
        def every_n_step_end(self, step, outputs):
            eval_results = self._estimator.evaluate(
                input_fn=input_fn_eval,
                metrics=eval_metrics,
                steps=None)
            print "Evaluation Result: %s" % eval_results
```

and the `eval_metrics` is is a list of `tf.contrib.metrics.streaming_sparse_recall_at_k`, `input_fn_eval` is a function to read TFRecord from file

but in the evaluation step, I always get the exception:

> File "...../tensorflow/contrib/learn/python/learn/graph_actions.py", line 610, in _eval_results_to_str
>     return ', '.join('%s = %s' % (k, v) for k, v in eval_results.items())

When i looks inside the evaluate method(Line 632- Line 779) in the source of [`graph_actions.py`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/graph_actions.py#L632)

```
def evaluate(graph,
             output_dir,
             checkpoint_path,
             eval_dict,
             update_op=None,
             global_step_tensor=None,
             supervisor_master='',
             log_every_steps=10,
             feed_fn=None,
             max_steps=None):
......
try:
      try:
        while (max_steps is None) or (step < max_steps):
          step += 1
          start_time = time.time()
          feed_dict = feed_fn() if feed_fn is not None else None
          if update_op is not None:
            session.run(update_op, feed_dict=feed_dict)
          else:
            eval_results = session.run(eval_dict, feed_dict=feed_dict)
            eval_step = step

          if step % log_every_steps == 0:
            if eval_step is None or step != eval_step:
              eval_results = session.run(eval_dict, feed_dict=feed_dict)
              eval_step = step
            duration = time.time() - start_time
            logging.info('Results after %d steps (%.3f sec/batch): %s.',
                         step, float(duration),
                         _eval_results_to_str(eval_results))
      finally:
        if eval_results is None or step != eval_step:
          eval_results = session.run(eval_dict, feed_dict=feed_dict)
          eval_step = step

......

 # Save summaries for this evaluation.
  _write_summary_results(output_dir, eval_results, current_global_step)

```

when my code call the function,   the `max step=2`(validation sample size=32, eval_batch_size=16) will be less than the default parameter( `log_every_steps=10`), therefore the code only can run the evaluation in the `finally` block, 

```
 ....
finally:
        if eval_results is None or step != eval_step:
          eval_results = session.run(eval_dict, feed_dict=feed_dict)
          eval_step = step
...
```

but for the metrics of t`tf.contrib.metrics.streaming_sparse_recall_at_k,` the `update_op` is not Not, which means the line

```
if update_op is not None
    session.run(update_op, feed_dict=feed_dict)
```

will alway be invoked, and the batched_data in the queue will be _empty_ when the code jump into `finally` block, therefore an exception of `OutOfRangeError` will be raised, and `eval_results` will be None when the code execute at line

```
_write_summary_results(output_dir, eval_results, current_global_step)
```

I think it is caused by the default value of `log_every_steps`, when the `_evaluate_model` method in `tf.contrib.learn.Estimator`  invoke the method in line [679](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L673),

```
eval_results, current_global_step = graph_actions.evaluate(
          graph=g,
          output_dir=eval_dir,
          checkpoint_path=checkpoint_path,
          eval_dict=eval_dict,
          update_op=update_op,
          global_step_tensor=global_step,
          supervisor_master=self._config.master,
          feed_fn=feed_fn,
          max_steps=steps)

```

it did not set the value of `log_every_steps` and  the default value of   `log_every_steps` will be 10, so the exception will be raised when the validation sample size is small (or total evaluation steps is less than 10)

When I set the hyper parameter `eval_batch_size` to a small value (eg. 2, which ensure total evaluation 32/2 = 16 > 10), my code is ok

**Comment by tatatodd** [comment]
Thanks for filing this bug @david-liu!  

Please add the below information to your bug, to help us investigate.  FYI this text is included in the default template for new issues when you click on the green "New issue" button at the top of this page.

Also - have you tried running on the latest 0.10 release?

```
### Environment info
Operating System:

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):

If installed from binary pip package, provide:

1. A link to the pip package you installed:
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.

If installed from source, provide 

1. The commit hash (`git rev-parse HEAD`)
2. The output of `bazel version`

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)

### What other attempted solutions have you tried?

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).
```

**Comment by david-liu** [comment]
Hello, Thank you response, I have run the code on the latest 0.10 release,

I think we can test with the source code of [A RETRIEVAL-BASED MODEL IN TENSORFLOW](https://github.com/dennybritz/chatbot-retrieval/ ) , after you download the data for google driver, 
and replace the validation.tfrecords with another validation tensorflow record with small size(2 validation sample,and I have created a r[epository](https://github.com/david-liu/validationTFRecord) to store this file, you can download it directly), and run the training process

python udc_train.py

you will get the error

**Comment by tatatodd** [comment]
@david-liu let's try to narrow down the problem first.  Can you try to fill out the full information that I requested?  E.g. the git commit hash you're installing from source, and other info.  Also please attach the full error log, including the full stack trace.

Note that log_every_steps just configures when logging occurs, and logging more or less often shouldn't affect your results.  Perhaps you should look into why you get an exception during logging, as that sounds like the root cause of your problem.

**Comment by mschulkind** [comment]
I'm having almost the exact same problem, but for a slightly different reason.

I'm using `graph_actions.evaluate` to evaluate my mode (through `BaseEstimator`, through `ValidationMonitor`), and I'm using an `input_fn` backed by a reader. I only want to run through my test data once, so I set `num_epochs=1`. I have 260 vectors in my test set, and `batch_size=100`, so as you'd expect, I get back 100, 100, 60 (because `allow_smaller_final_batch=True`), and then the next time  I get an `OutOfRangeError` which looks like it reproduces almost exactly the same problem.

I'm running about 0.10. I haven't had the time to try against master, but it seems like the relevant code hasn't changed there.

The line that causes the exception is:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/graph_actions.py#L754
```
      finally:
        if eval_results is None or step != eval_step:
          eval_results = session.run(eval_dict, feed_dict=feed_dict) # <---- This is the culprit.
          eval_step = step
        # Stop session first, before queue runners.
        session.close()
```

This means that `eval_results` never gets set, and the session doesn't even get closed.

The exception handler catches it just fine at:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/graph_actions.py#L768
```
    except errors.OutOfRangeError as e:
      if max_steps is None:
        logging.info('Input queue is exhausted.')
      else:
        logging.warn('Input queue is exhausted: %s.', e)
```

...but this means that the call at:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/graph_actions.py#L781
```
  _write_summary_results(output_dir, eval_results, current_global_step)
```
gets `None` for `eval_results`, which then causes problems at...
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/graph_actions.py#L612
```
  return ', '.join('%s = %s' % (k, v) for k, v in eval_results.items())
```
because you can't call `.items()` on `None`.

**Comment by mschulkind** [comment]
This looks like the guilty commit:
https://github.com/tensorflow/tensorflow/commit/54cf1600b98127d5076def3c567cda1e087c1761

Basically, it seems like if you ever hit either of the 2 except clauses because your `eval_dict` throws the matching exceptions, you're going to then get an exception from `_write_summary_results`.

**Comment by tatatodd** [comment]
@ilblackdragon can you comment on this?  Thanks!

**Comment by drpngx** [comment]
Is this still current?

**Comment by drpngx** [comment]
Closing due to inactivity. Feel free to re-open if you would like us to look again.
